src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -60,58 +65,340 @@ public class ResponseCachingMiddleware
+                    foreach (var varyKey in context.KeyProvider.CreateLookupVaryKeys(context))
+                    {
+                        cacheEntry = context.Cache.Get(varyKey);
+
+                        if (cacheEntry is CachedResponse && await TryServeCachedResponseAsync(context, cacheEntry as CachedResponse))
+                        {
+                            return true;
+                        }
                     }
-                    finally
+                }
+
+                if (cacheEntry is CachedResponse && await TryServeCachedResponseAsync(context, cacheEntry as CachedResponse))
+                {
+                    return true;
+                }
+            }
+
+
+            if (context.RequestCacheControlHeaderValue.OnlyIfCached)
+            {
+                context.HttpContext.Response.StatusCode = StatusCodes.Status504GatewayTimeout;
+                return true;
+            }
+
+            return false;
+        }
+
+        internal static bool ConditionalRequestSatisfied(ResponseCachingContext context)
+        {
+            var cachedResponseHeaders = context.CachedResponseHeaders;
+            var ifNoneMatchHeader = context.TypedRequestHeaders.IfNoneMatch;
+
+            if (ifNoneMatchHeader != null)
+            {
+                if (ifNoneMatchHeader.Count == 1 && ifNoneMatchHeader[0].Equals(EntityTagHeaderValue.Any))
+                {
+                    return true;
+                }
+
+                if (cachedResponseHeaders.ETag != null)
+                {
+                    foreach (var tag in ifNoneMatchHeader)
                     {
-                        cachingContext.UnshimResponseStream();
+                        if (cachedResponseHeaders.ETag.Compare(tag, useStrongComparison: true))
+                        {
+                            return true;
+                        }
                     }
                 }
+            }
+            else if ((cachedResponseHeaders.LastModified ?? cachedResponseHeaders.Date) <= context.TypedRequestHeaders.IfUnmodifiedSince)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        internal static void FinalizeCachingHeaders(ResponseCachingContext context)
+        {
+            if (context.CacheabilityValidator.IsResponseCacheable(context))
+            {
+                context.ShouldCacheResponse = true;
+                context.StorageBaseKey = context.KeyProvider.CreateStorageBaseKey(context);
+
+                // Create the cache entry now
+                var response = context.HttpContext.Response;
+                var varyHeaderValue = response.Headers[HeaderNames.Vary];
+                var varyParamsValue = context.HttpContext.GetResponseCachingFeature()?.VaryParams ?? StringValues.Empty;
+                context.CachedResponseValidFor = context.ResponseCacheControlHeaderValue.SharedMaxAge
+                    ?? context.ResponseCacheControlHeaderValue.MaxAge
```

Comment:
Operator on the previous line: `context.ResponseCacheControlHeaderValue.SharedMaxAge ?? ...`